index.rst

```diff
@@ -48,14 +48,177 @@
+
+The first ensures that workflow completion coincides with registry updates but could lead to an arbitrary number of these jobs attempting to sync up with the main registry simultaneously.
+The second decouples workflow completion from registry updating and allows a rate-limited number of updates to occur in parallel.
+This would lead to a situation where a workflow can complete but the registry it out of date for an indeterminate period of time and would delay submission of workflows that depend on the results.
+Were that to happen though, it would be indicative that letting each workflow attempt the sync up directly at the end would be risky.
+Using a queue also completely removes registry credentials from workflow execution since the queued updates would be running in a completely different environment.
+
+The synchronization must check that a corresponding dataset was written to the object store since it is possible for a workflow to partially complete and synchronization should (optionally?) happen even on failure since that can allow resubmission of the workflow with different configurations and also allow investigation of the intermediate products.
+
+The limited registry approach still requires that the butler is interacting with an object store and would require the use of signed URLs if the datasets were being written to their final location in the shared datastore.
+If the signing client becomes a bottleneck an alternative could be to use a temporary bucket specifically for that workflow.
+
+The synchronization job would therefore also need to move the files from the temporary location to the final location.
+
+
+Summary
+=======
+
+Whilst the job-level pooling and externalized approaches will help with registry contention, by far the safest approach, and in the long term the simplest, is to adopt the limited read-only registry.
+This removes any worries about merging the outputs of multiple jobs and lets us leverage the knowledge already known to the graph builder.
+
+Task Breakdown
+--------------
+
+The initial prototype implementation will not include URL signing since all Data Preview 0.2 workflow executions will be run by staff.
+
+An initial break down of work would then be:
+
+* Add facility to pipetask command line to create a minimalist SQLite registry from the constructed quantum graph.
+  Initially this registry could ignore the datastore table completely.
+* Create a new ``Datastore`` subclass that no longer queries registry on ``Datastore.get()`` but instead assumes that the active datastore configuration (read from the user-supplied Butler configuration) would give the correct answer (something that can not be relied on in general but can be relied on in this limited context) for file template and formatter class.
+  Also modify ``Datastore.put()`` such that the registry is never written to (this could be implemented as a new limited datastore registry manager class).
```

Comment:
I think it might be better if we just let `Datastore.put` update the registry database here as it usually does.  Those changes will not survive the job, because we won't be transferring that copy of the SQLite file to downstream jobs, but it provides a bit of diagnostic information about what incomplete jobs accomplished, and that might be useful (especially in the local, developer-mode version of this I'm imagining).

Reply:
Doesn't this require that we merge datastore registries again though? When you do a coadd you will now have different input sqlite registries from the jobs that created the calexps and warps. You can always do the write but the read has to have everything regardless. I was envisaging a manager class that compares the write with what was in the table and says "fine" if they match (modulo size and checksum fields).

Reply:
I was imagining that we'd keep the `Datastore.get` logic you proposed that didn't expect the datastore records to be populated, for at least one of the chained datastores.  I think the most general case is that we'd chain (in `get` priority order):
1. A normal local-scratch filesystem datastore (first priority for `get`, also `puts` at least sometimes);
2. An object-store datastore that uses internal records for `get` and `put`, as usual.
3. An object-store datastore that that uses configuration to predict URIs and formatters on `get`.  Never needs to `put`.  Actually has the same root as (2).

Lots of ways to potentially merge (2) and (3), and sometimes we may not want (1).  But hopefully that explains what I had in mind.

Reply:
Okay, so we do write to registry on put but the sqlite file is never used for gets, which means that for debugging you find the one that was harvested by the workflow manager?  I don't think it matters whether we are doing local scratch filesystem read/writes or object store does it? A chained datastore might actually be a bit confusing with the datastore tweaking we are talking about (one with a local write for caching and a put to object store) but probably could be made to work. I thought we were arguing against doing local writes that would be used by the jobs and handled by the workflow manager because that would involve lots of node to node copying to a shared storage managed by the workflow manager and if many jobs are running that storage becomes large and it's easy to manage the object store directly.

Reply:
> Okay, so we do write to registry on put but the sqlite file is never used for gets, which means that for debugging you find the one that was harvested by the workflow manager?

Yup, that is what I'm proposing.

> I don't think it matters whether we are doing local scratch filesystem read/writes or object store does it?
> ...
> I thought we were arguing against doing local writes that would be used by the jobs and handled by the workflow manager

Yeah, I only added local scratch here as something we _might_ do for within-a-job I/O optimization (i.e. when a job is a pretty sizeable graph itself).  I was actually trying to avoid causing confusion by writing a list of datasets that _didn't_ include that, but I think I just created confusion by making it sound like that's related to how we read from and write to the object store.

And I'm agnostic as to whether we handle those object-store reads and writes via a chained datastore or not - the important point I wanted to make was just that we can write datastore records to the local database as long as we don't need them to be able to read from that datastore in another job.